### PR TITLE
Improve loggers

### PR DESCRIPTION
**Pull request by blacksponge**
## Purpose

Following the discussion at https://github.com/zonemaster/zonemaster-backend/pull/959#pullrequestreview-898947996 I decided to do a dedicated PR for logging.

## Context

* Replace part of #959
* Maybe address partially #214

## Changes
* Drop Log::Dispatcher
* Implementation of log levels now corresponds to the documentation
* Add support for structured logging (JSON) using context and extra parameters
  * `ZM_BACKEND_RPCAPI_LOGJSON=1` for RPCAPI
  * `--logjson` for test agent
* Use same init function for logger from RPCAPI script and test agent

## How to test this PR
* Structured logging

`ZM_BACKEND_RPCAPI_LOGJSON=1 ZM_BACKEND_RPCAPI_LOGLEVEL=debug starman --preload-app script/zonemaster_backend_rpcapi.psgi  2>&1 | jq -crR '. as $raw | try fromjson catch $raw'`
```json
{"category":"Zonemaster::Backend::Config","level":"notice","message":"Loading config: /home/gbm/workspace/zonemaster/backend_config.ini","pid":91858,"timestamp":"2022-03-28T16:11:45Z"}
{"category":"Plack::Sandbox::_2fhome_2fgbm_2fworkspace_2fzonemaster_2fzonemaster_2dbackend_2fscript_2fzonemaster_backend_rpcapi_2epsgi","level":"info","message":"Enabling add_api_user method","pid":91858,"timestamp":"2022-03-28T16:11:45Z"}
{"category":"Plack::Sandbox::_2fhome_2fgbm_2fworkspace_2fzonemaster_2fzonemaster_2dbackend_2fscript_2fzonemaster_backend_rpcapi_2epsgi","level":"info","message":"Enabling add_batch_job method","pid":91858,"timestamp":"2022-03-28T16:11:45Z"}
{"category":"Zonemaster::Backend::DB","level":"notice","message":"Connecting to database 'DBI:Pg:dbname=travis_zonemaster;host=127.0.0.1;port=5432' as user 'travis_zonemaster'","pid":91858,"timestamp":"2022-03-28T16:11:49Z"}
{"category":"Zonemaster::Backend::RPCAPI","code":-32001,"data":{"username":"toto"},"error":"Zonemaster::Backend::Error::PermissionDenied","level":"info","message":"User not authorized to use batch mode (code -32001). Context: {'username' => 'toto'}","pid":91858,"timestamp":"2022-03-28T16:11:49Z"}
```
`perl script/zonemaster_backend_testagent --logfile=- --outfile=/tmp/zonemaster_testagent.out --loglevel=debug foreground --logjson | jq -c`
```json
{"category":"main","level":"notice","message":"Daemon spawned","pid":91993,"timestamp":"2022-03-28T16:13:44Z"}
{"category":"Zonemaster::Backend::Config","level":"notice","message":"Loading config: /home/gbm/workspace/zonemaster/backend_config.ini","pid":91993,"timestamp":"2022-03-28T16:13:44Z"}
{"category":"Zonemaster::Backend::DB","level":"notice","message":"Connecting to database 'DBI:Pg:dbname=travis_zonemaster;host=10.10.86.42;port=5432' as user 'travis_zonemaster'","pid":91993,"timestamp":"2022-03-28T16:13:45Z"}
{"category":"main","level":"error","message":"DBI connect('dbname=travis_zonemaster;host=10.10.86.42;port=5432','travis_zonemaster',...) failed: n'a pas pu se connecter au serveur : Connexion refusée\n\tLe serveur est-il actif sur l'hôte « 10.10.86.42 » et accepte-t-il les connexions\n\tTCP/IP sur le port 5432 ? at /home/gbm/workspace/zonemaster/zonemaster-backend/lib/Zonemaster/Backend/DB.pm line 96.\n","pid":91993,"timestamp":"2022-03-28T16:13:45Z"
```
* Log::Any log levels work correctly

`perl script/zonemaster_backend_testagent --logfile=- --outfile=/tmp/zonemaster_testagent.out --loglevel=trace foreground`
```
2022-03-28T16:19:35Z [92354] [NOTICE] [main] Daemon spawned
2022-03-28T16:19:35Z [92354] [NOTICE] [Zonemaster::Backend::Config] Loading config: /home/gbm/workspace/zonemaster/backend_config.ini
```

### CODE REVIEW

**Comment by matsduf** [comment]
>     * Add support for structured logging (JSON) using context and extra parameters
>       
>       * `ZM_BACKEND_RPCAPI_LOGJSON=1` for RPCAPI
>       * `--logjson` for test agent

Why not in the same way?

**Comment by blacksponge** [comment]
> What documentation are you referring to?

https://github.com/zonemaster/zonemaster-backend/blob/b0e46a8477c7da5c150b9babf6acb110fab8143c/script/zonemaster_backend_testagent#L295-L300

> Why not in the same way?

Logger parameters are currently passed in the command line in the test agent and using environment variable for the RPCAPI, I tried to keep things coherent regarding their respective implementation.

**Comment by blacksponge** [comment]
> Maybe we should open an issue with Log::Any to improve the situation?
> Is JSON-output the sole reason why you went the way of creating a new adapter, or are the existing ones lacking in some other way?

No, the JSON output is a nice to have that I wanted to implement for a while and I took the opportunity to add it here as I was working on logging. The default adapters are seriously lacking some core features to my eyes, there is no support for structured logging (so we can not add context) but more importantly we can not configure the log format to add the PID or the log level for instance and the log format isn't even uniform between the different standard adapters:
* `my $msg = sprintf( "[%s] %s\n", scalar(localtime), $text );` for [`Log::Any::Adapter::File`](https://metacpan.org/module/Log::Any::Adapter::File/source)
* `print STDOUT "$text\n";` for [`Log::Any::Adapter::Stdout`](https://metacpan.org/module/Log::Any::Adapter::Stdout/source)

Maybe we could open an issue, but in the mean time I'd like to have a working and uniform logger (and json support). If someone has a better way to do this I am happy to change the implementation.

File: lib/Zonemaster/Backend/Log.pm [hunk 1]
```diff
@@ -0,0 +1,118 @@
...
+
+my $trace_level = Log::Any::Adapter::Util::numeric_level('trace');
+
+sub init {
+    my ($self) = @_;
+
+    if ( exists $self->{log_level} && $self->{log_level} =~ /\D/ ) {
+        my $numeric_level = Log::Any::Adapter::Util::numeric_level( $self->{log_level} );
+        if ( !defined($numeric_level) ) {
+            die "Error: Unrecognized log level " . $self->{log_level} . "\n";
```
[the remark this answers is not in the document]
Comment (blacksponge) [reply lib/Zonemaster/Backend/Log.pm:19]: I admit I don't know all the subtlety, I will change that

File: lib/Zonemaster/Backend/Log.pm [hunk 2]
```diff
@@ -0,0 +1,118 @@
...
+        level => $level,
+        category => $category,
+        message => $string
+    );
+
+    for my $item ( @items ) {
+        if (ref($item) eq 'HASH') {
+            if (exists $item->{context}) {
+                $item = $item->{context};
+            }
```
[the remark this answers is not in the document]
Comment (blacksponge) [reply lib/Zonemaster/Backend/Log.pm:97]: I thought that `context` needed some special case but I was wrong, I will remove that.

File: lib/Zonemaster/Backend/Log.pm [hunk 3]
```diff
@@ -0,0 +1,118 @@
...
+    if (exists $log_params->{pid}) {
+        $msg .= sprintf "[%d] ", $log_params->{pid};
+        delete $log_params->{pid};
+    }
+    $msg .= sprintf "[%s] [%s] %s", uc $log_params->{level}, $log_params->{category}, $log_params->{message};
+    delete $log_params->{level};
+    delete $log_params->{message};
+    delete $log_params->{category};
+
+    return $msg
```
[the remark this answers is not in the document]
Comment (blacksponge) [reply lib/Zonemaster/Backend/Log.pm:65]: I wasn't sure whether to add it to the text logger too, but it is probably a good idea indeed

File: lib/Zonemaster/Backend/Log.pm [hunk 4]
```diff
@@ -0,0 +1,118 @@
...
+
+    my $js = JSON::PP->new;
+    $js->canonical( 1 );
+
+    return $js->encode( $log_params );
+}
+
+
+sub structured {
+    my ($self, $level, $category, $string, @items) = @_;
```
[the remark this answers is not in the document]
Comment (blacksponge) [reply lib/Zonemaster/Backend/Log.pm:80]: Actually the context is also passed in this list so if we want to limit it, it should be a maximum of 2. But we could have an incoherent behavior when no context is used as we would be accepting two hashrefs and not one, so I prefer not having a strict limit here. Also the implementation in Log::Any do allow for more than one to be provided.

**Review by matsduf — APPROVED** [review]

File: script/zonemaster_backend_testagent [hunk 5]
```diff
@@ -60,56 +61,13 @@ $opt_outfile //= '/var/log/zonemaster/zonemaster_backend_testagent.out';
...
-        ]
-    );
-}
+Log::Any::Adapter->set(
+    '+Zonemaster::Backend::Log',
+    log_level => $loglevel,
+    json => $logjson,
+    file => $logfile,
+);
+$log->context->{pid} = $PID;
```
Comment (mattias-p) [inline script/zonemaster_backend_testagent:70]: Here we assign the PID of the master process, right? Will all workers be logged with the same pid?

File: script/zonemaster_backend_rpcapi.psgi [hunk 6]
```diff
@@ -30,34 +28,13 @@ use Zonemaster::Backend::Metrics;
...
-                callbacks => sub {
-                    my %args = @_;
-                    $args{message} = sprintf "%s [%d] %s - %s", strftime( "%FT%TZ", gmtime ), $PID, uc $args{level}, $args{message};
-                },
-            ],
-        ]
-    ),
+    '+Zonemaster::Backend::Log',
+    log_level => lc $ENV{ZM_BACKEND_RPCAPI_LOGLEVEL},
+    json => $ENV{ZM_BACKEND_RPCAPI_LOGJSON},
```
Comment (mattias-p) [inline script/zonemaster_backend_rpcapi.psgi:34]: I guess these aren't mentioned anywhere. Maybe we should mention them somewhere so people can find them? We already have docs/Configuration.md for the config file. Maybe we should have one for environment variables as well?

File: lib/Zonemaster/Backend/Log.pm [hunk 7]
```diff
@@ -0,0 +1,126 @@
...
+        my $numeric_level = Log::Any::Adapter::Util::numeric_level( $self->{log_level} );
+        if ( !defined($numeric_level) ) {
+            croak "Error: Unrecognized log level " . $self->{log_level} . "\n";
+        }
+        $self->{log_level} = $numeric_level;
+    }
+
+    if ( !defined $self->{log_level} ) {
+        $self->{log_level} = $trace_level;
+    }
```
[the remark this answers is not in the document]
Comment (blacksponge) [reply lib/Zonemaster/Backend/Log.pm]: Done.

File: lib/Zonemaster/Backend/Log.pm [hunk 8]
```diff
@@ -0,0 +1,126 @@
...
+        if ( $self->{stderr} ) {
+            open( $fd, '>&', \*STDERR ) or die "Can't dup STDERR: $!";
+        } else {
+            open( $fd, '>&', \*STDOUT ) or die "Can't dup STDOUT: $!";
+        }
+        my $handle = IO::Handle->new_from_fd( $fd, "w" ) or die "Can't fdopen duplicated STDOUT: $!";
+    } else {
+        open( $fd, '>>', $self->{file} ) or die "Can't open log file: $!";
+    }
+    $self->{handle} = IO::Handle->new_from_fd( $fd, "w" ) or die "Can't fdopen file: $!";
```
[the remark this answers is not in the document]
Comment (blacksponge) [reply lib/Zonemaster/Backend/Log.pm]: Done

File: lib/Zonemaster/Backend/Log.pm [hunk 9]
```diff
@@ -0,0 +1,126 @@
...
+    }
+
+    my $fd;
+    if ( !exists $self->{file} || $self->{file} eq '-') {
+        if ( $self->{stderr} ) {
+            open( $fd, '>&', \*STDERR ) or die "Can't dup STDERR: $!";
+        } else {
+            open( $fd, '>&', \*STDOUT ) or die "Can't dup STDOUT: $!";
+        }
+        my $handle = IO::Handle->new_from_fd( $fd, "w" ) or die "Can't fdopen duplicated STDOUT: $!";
```
[the remark this answers is not in the document]
Comment (blacksponge) [reply lib/Zonemaster/Backend/Log.pm]: None, probably the result of bad copy/paste, fixed.

File: lib/Zonemaster/Backend/Log.pm [hunk 10]
```diff
@@ -0,0 +1,126 @@
...
+            $self->{formatter} = \&format_text;
+        }
+    }
+}
+
+sub format_text {
+    my ($self, $log_params) = @_;
+    my $msg;
+    $msg .= sprintf "%s ", $log_params->{timestamp};
+    delete $log_params->{timestamp};
```
[the remark this answers is not in the document]
Comment (blacksponge) [reply lib/Zonemaster/Backend/Log.pm:58]: Thanks, for the tips, done.

File: lib/Zonemaster/Backend/Log.pm [hunk 11]
```diff
@@ -0,0 +1,126 @@
...
+    if (exists $log_params->{pid}) {
+        $msg .= sprintf "[%d] ", $log_params->{pid};
+        delete $log_params->{pid};
+    }
+    $msg .= sprintf "[%s] [%s] %s", uc $log_params->{level}, $log_params->{category}, $log_params->{message};
+    delete $log_params->{level};
+    delete $log_params->{message};
+    delete $log_params->{category};
+
+    if ( keys %$log_params ) {
```
[the remark this answers is not in the document]
Comment (blacksponge) [reply lib/Zonemaster/Backend/Log.pm]: Done.

File: script/zonemaster_backend_testagent (hunk 5 above)
Comment (blacksponge) [reply script/zonemaster_backend_testagent:70]: Indeed it was always the master process pid, fixed.

File: script/zonemaster_backend_rpcapi.psgi [hunk 12]
```diff
@@ -30,34 +28,13 @@ use Zonemaster::Backend::Metrics;
...
+    stderr => 1
 );
+$log->context->{pid} = $PID;
```
[the remark this answers is not in the document]
Comment (blacksponge) [reply script/zonemaster_backend_rpcapi.psgi]: Fixed.

File: lib/Zonemaster/Backend/Log.pm [hunk 13]
```diff
@@ -0,0 +1,126 @@
...
+        timestamp => strftime( "%FT%TZ", gmtime ),
+        level => $level,
+        category => $category,
+        message => $string
+    );
+
+    for my $item ( @items ) {
+        if (ref($item) eq 'HASH') {
+            for my $key (keys %$item) {
+                $log_params{$key} = $item->{$key};
```
[the remark this answers is not in the document]
Comment (blacksponge) [reply lib/Zonemaster/Backend/Log.pm:106]: I did but I considered it too extra work for an interface only used internally. Also what should be the behavior? Die / log warn?

**Comment by matsduf** [comment]
The new "docs/Configuration.md" is a good addition. This is outside the scope of this PR, but we also use environment variable `ZONEMASTER_BACKEND_CONFIG_FILE` in `Zonemaster::Backend::Config` which should be added to the new document. Now or in a later PR.

**Review by matsduf — APPROVED** [review]

File: lib/Zonemaster/Backend/Log.pm (hunk 13 above)
[the remark this answers is not in the document]
Comment (mattias-p) [reply lib/Zonemaster/Backend/Log.pm:106]: Good question. I guess a call to `warn()` would be the reasonable choice. But maybe it isn't worth the effort. I'll leave it up to you.

File: docs/EnvironmentVariables.md [hunk 14]
```diff
@@ -0,0 +1,18 @@
...
+  * `debug`
+  * `info` (also accepted: `inform`)
+  * `notice`
+  * `warning` (also accepted: `warn`)
+  * `error` (also accepted: `err`)
+  * `critical` (also accepted: `crit`, `fatal`)
+  * `alert`
+  * `emergency`
+* `ZM_BACKEND_RPCAPI_LOGJSON`: Setting it to any value will configure the
+  logger to log in JSON format, undefined by default.
```
Comment (mattias-p) [inline docs/EnvironmentVariables.md:18]: The emtpy string and the string `0` are also interpreted as false.

**Comment by blacksponge** [comment]
> The new "docs/Configuration.md" is a good addition. This is outside the scope of this PR, but we also use environment variable ZONEMASTER_BACKEND_CONFIG_FILE in Zonemaster::Backend::Config which should be added to the new document. Now or in a later PR.

That's right, thanks for pointing it out, I will add this one in a subsequent PR.

File: docs/EnvironmentVariables.md (hunk 14 above)
Comment (blacksponge) [reply docs/EnvironmentVariables.md]: Should be better now

**Review by mattias-p — APPROVED** [review]

**Pull request merged**